refactor(cameras): cameras implementations + tests improvements

imstevenpmwork opened the pull request:
This PR refactors the `cameras` module, improving its structure, reliability, and test coverage:

#### Key Changes

* Abstract Base Class (`Camera`): Standardizes core parameters and methods for all camera implementations.
* Enhanced Camera Implementations: `OpenCVCamera`, `RealSenseCamera`
* Improved Utilities & Tests:
   - Refines `find_cameras` for better device discovery.
   - Adds unit tests for OpenCV & RealSense cameras.
   - Migrates heavy test assets to **Git LFS** and removes unused files.
* Dependencies and Docs:
   - Adds `pyrealsense2` for macOS support.
   - Enhances inline docs.

#### Why This Matters
- Consistency: A unified base class simplifies adding future camera types.
- Reliability: Better depth handling, warmup, and tests prevent common issues.
- Maintainability: Cleaner code, docs, and tests ensure long-term stability.

#### Testing
* Unit Tests: Validated core functionality for both camera types.  
* CI: Confirmed compatibility on Linux/macOS.
* Manual HW Checks: Verified code with RealSense D435 and multiple webcams.

aliberts commented on lerobot/common/robots/config.py:
```diff
@@ -26,6 +26,17 @@ class RobotConfig(draccus.ChoiceRegistry, abc.ABC):
     calibration_dir: Path | None = None
 
+    def __post_init__(self):
+        if hasattr(self, "cameras"):
+            cameras = self.cameras
+            if cameras:
+                for cam_name, cam_config in cameras.items():
+                    for attr in ["width", "height", "fps"]:
+                        if getattr(cam_config, attr) is None:
+                            raise ValueError(
+                                f"Camera config for '{cam_name}' has None value for required attribute '{attr}'"
+                            )
```
```suggestion
        if hasattr(self, "cameras") and self.cameras:
            for name, config in self.cameras.items():
                for attr in ["width", "height", "fps"]:
                    if getattr(config, attr) is None:
                        raise ValueError(
                            f"Specifying '{attr}' is required for the camera to be used in a robot"
                            )
```
Also did you make sure this is run? We don't need to `super().__post_init__()` in the subclasses?

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/a36536dc078cf7f04ccb2b302cb458506d3634af

aliberts commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,357 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
Can you add a short docstring description (1 sentence) and an example of how to run it here? (similar to what we have in the other scripts)

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/41179d0996122b912e10e9989f8a05e91faca68b

aliberts commented on lerobot/teleoperate.py:
```diff
@@ -38,7 +38,6 @@
 import numpy as np
 import rerun as rr
 
-from lerobot.common.cameras import intel, opencv  # noqa: F401
```
But then the cameras won't work, no?

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/commit/525cd68e62c61667860018d354b6c32d83874a78

aliberts commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,190 @@
+from lerobot.common.errors import DeviceAlreadyConnectedError, DeviceNotConnectedError
+
+# NOTE(Steven): more tests + assertions?
+TEST_ARTIFACTS_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "artifacts", "cameras")
+DEFAULT_PNG_FILE_PATH = os.path.join(TEST_ARTIFACTS_DIR, "fakecam_sd_160x120.png")
+TEST_IMAGE_PATHS = [
+    os.path.join(TEST_ARTIFACTS_DIR, "fakecam_sd_160x120.png"),
+    os.path.join(TEST_ARTIFACTS_DIR, "fakecam_hd_320x180.png"),
+    os.path.join(TEST_ARTIFACTS_DIR, "fakecam_fullhd_480x270.png"),
+    os.path.join(TEST_ARTIFACTS_DIR, "fakecam_square_128x128.png"),
+]
```
Let's use Pathlib.
Also, let's use simpler name for these artifacts, e.g. just `image_160x120` instead of `fakecam_sd_160x120`

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/41179d0996122b912e10e9989f8a05e91faca68b

aliberts commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,190 @@
+        _ = camera.disconnect()
+
+
+@pytest.mark.parametrize("index_or_path", TEST_IMAGE_PATHS)
```
same

aliberts reviewed (changes requested):
Thank you for this refactor, here's my first round of review.

I have reviewed realsense more extensively than opencv, but a lot of my comments on the former should also be applied to the latter.

Overall good progress but I find there are too many logging statements that are adding noise and not much help. I've added suggestion to remove some of them but I didn't do it for all of them as there were many, let's clean that up a bit.

Cadene commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+except (ImportError, ModuleNotFoundError):
+    pytest.skip("pyrealsense2 not available", allow_module_level=True)
+
+TEST_ARTIFACTS_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "artifacts", "cameras")
+BAG_FILE_PATH = os.path.join(TEST_ARTIFACTS_DIR, "test_rs.bag")
```
Let's CTRL+F `os.path.join` and replace by `Path`

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/41179d0996122b912e10e9989f8a05e91faca68b

Cadene reviewed (commented):
Nice first version! Here are my comments ;)

aliberts commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,357 @@
+from lerobot.common.cameras.opencv.camera_opencv import OpenCVCamera
+from lerobot.common.cameras.opencv.configuration_opencv import OpenCVCameraConfig
+
+logger = logging.getLogger(__name__)
```
This so that someone using lerobot as a dependency doesn't get polluted by print statements that are out of his/her control but rather to give them control over it with the `logging` tools. This is standard practice for a library.

Indeed, we need to init_logging in our own scripts if we want to see the statements.

imstevenpmwork replied:
Can we resolve this conversation?

imstevenpmwork commented on lerobot/common/cameras/camera.py:
```diff
@@ -1,25 +1,49 @@
+        self.width: int | None = config.width
+        self.height: int | None = config.height
+
+    @property
     @abc.abstractmethod
```
`Using deprecated decorator abc.abstractproperty(). The class "abstractproperty" is deprecated. Use 'property' with 'abstractmethod' instead.` Do we still want to use `@abstractproperty` despite it being deprecated?

aliberts replied:
Cool, didn't know that! I though this was only the case with abstractclassmethod and abstractstaticmethod.
I've updated my branch accordingly [8ab2227](https://github.com/huggingface/lerobot/pull/777/commits/8ab22271481b4ab65ad8037eb16e1ebe6c1a5ed2)

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+            self.rs_profile = self.rs_pipeline.start(rs_config)
+            logger.debug(f"Successfully started pipeline for camera {self.serial_number}.")
+        except RuntimeError as e:
+            self.rs_profile = None
+            self.rs_pipeline = None
```
While not required, an user might handle camera connections like this:
```python
try:
    camera.connect()
    ...
finally:
    camera.disconnect()
```

In `disconnect()`, we safely call `self.rs_pipeline.stop()` only if `rs_pipeline` exists. To prevent errors from invalid objects, we explicitly set `rs_pipeline = None` when something went wrong. This ensures clean shutdowns without raising exceptions.
We could remove it tho

aliberts replied:
I see, thanks!

imstevenpmwork commented on lerobot/common/cameras/camera.py:
```diff
@@ -1,25 +1,49 @@
+from .configs import CameraConfig, ColorMode
+
 
 class Camera(abc.ABC):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/41179d0996122b912e10e9989f8a05e91faca68b

imstevenpmwork commented on lerobot/common/cameras/intel/__init__.py:
Addressed in commit: https://github.com/huggingface/lerobot/commit/9812f2db09f2c1f04bad89e41b001cea287fff86

imstevenpmwork commented on lerobot/calibrate.py:
```diff
@@ -31,7 +31,6 @@
 
 import draccus
 
-from lerobot.common.cameras import intel, opencv  # noqa: F401
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/525cd68e62c61667860018d354b6c32d83874a78

imstevenpmwork commented on lerobot/common/cameras/utils.py:
```diff
@@ -19,3 +44,30 @@ def make_cameras_from_configs(camera_configs: dict[str, CameraConfig]) -> dict[s
+        Cv2Rotation.ROTATE_270: cv2.ROTATE_90_COUNTERCLOCKWISE,
+        Cv2Rotation.ROTATE_90: cv2.ROTATE_90_CLOCKWISE,
+        Cv2Rotation.ROTATE_180: cv2.ROTATE_180,
+    }.get(rotation)
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/ab7565a666024c3ed276c93ec317f4eb79c44df0

imstevenpmwork commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+        Cv2Rotation.ROTATE_180,
+        Cv2Rotation.ROTATE_270,
+    ],
+)
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/efd2849184b1b005ddd91c270655caac8640f49e

imstevenpmwork commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,190 @@
+        _ = camera.async_read()
+
+
+@pytest.mark.parametrize("index_or_path", TEST_IMAGE_PATHS)
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/efd2849184b1b005ddd91c270655caac8640f49e

imstevenpmwork commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,190 @@
+        camera.connect(do_warmup_read=False)
+
+
+@pytest.mark.parametrize("index_or_path", TEST_IMAGE_PATHS)
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/efd2849184b1b005ddd91c270655caac8640f49e

imstevenpmwork commented on lerobot/record.py:
```diff
@@ -44,8 +44,6 @@
 
 from lerobot.common.cameras import (  # noqa: F401
     CameraConfig,  # noqa: F401
-    intel,
-    opencv,
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/525cd68e62c61667860018d354b6c32d83874a78

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+    def is_connected(self) -> bool:
+        """Checks if the camera is currently connected and opened."""
+        return isinstance(self.videocapture_camera, cv2.VideoCapture) and self.videocapture_camera.isOpened()
 
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/99ea24d6a3cb8e06c60e1669d2c29ff4c3e3714f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        depth_camera = RealSenseCamera(custom_config)
+        try:
+            depth_camera.connect()
+            color_image, depth_map = depth_camera.read() # Returns tuple
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/b3dafcfb077a489816ee4b271abe472840663dc7

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        elif config.serial_number is not None:
+            self.serial_number = str(config.serial_number)
+        else:
+            raise ValueError("RealSenseCameraConfig must provide either 'serial_number' or 'name'.")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/18b56e15336e5c133ba58bc33bf65fcddf246e73

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        Initializes the RealSense pipeline, configures the required streams (color
+        and optionally depth), starts the pipeline, and validates the actual stream settings.
+
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/eb2d967c85b996784e83bc48b5b9487c8b2f70d1
Although I added in the docs of the `Camera` class

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+                self.prerotated_width, self.prerotated_height = actual_width, actual_height
+            logger.info(f"Capture width set to camera default: {self.width}.")
+            logger.info(f"Capture height set to camera default: {self.height}.")
+            return
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/39f908e9db68e82a8e667a3ff4373dd06c587a99

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+                f"Error getting frame data from queue for camera {self.serial_number}: {e}"
+            ) from e
+
+    def _shutdown_read_thread(self):
```
Addressed in commit (the move): https://github.com/huggingface/lerobot/pull/1108/commits/a65321017358b2a94840ea985239ac2abdf7d0b6

imstevenpmwork commented on lerobot/common/cameras/intel/configuration_realsense.py:
```diff
@@ -14,58 +14,74 @@
-    mock: bool = False
+    rotation: Cv2Rotation = Cv2Rotation.NO_ROTATION  # NOTE(Steven): Check if draccus can parse to an enum
 
     def __post_init__(self):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/91a5fafdaebcf6550a53378f1e16800d637a319f

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/91a5fafdaebcf6550a53378f1e16800d637a319f

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
-        self.capture_height = round(actual_height)
-        self.is_connected = True
+    @staticmethod
+    def find_cameras(
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/33b85590607e2ee39d43f83bf3d5b705ea8433cf

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+            logger.exception(f"Unexpected error getting frame from queue for {self}: {e}")
+            raise RuntimeError(f"Error getting frame from queue for camera {self.index_or_path}: {e}") from e
+
+    def _shutdown_read_thread(self):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/a65321017358b2a94840ea985239ac2abdf7d0b6

imstevenpmwork commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,357 @@
+    )
+    subparsers = parser.add_subparsers(dest="command", help="Available commands")
+
+    # List cameras command
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/41179d0996122b912e10e9989f8a05e91faca68b#diff-b9bfa304658b8552d5d648c0bcbf565a2e68dc51c1df5e2d5626b2b2e765c3fb

imstevenpmwork commented on lerobot/common/cameras/camera.py:
```diff
@@ -1,25 +1,49 @@
 
     @abc.abstractmethod
-    def read(self, temporary_color: str | None = None) -> np.ndarray:
+    def connect(self, do_warmup_read: bool = True) -> None:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on lerobot/common/cameras/__init__.py:
```diff
@@ -1,3 +1,17 @@
+# limitations under the License.
+
 from .camera import Camera
 from .configs import CameraConfig
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

aliberts commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        if do_warmup_read:
+            logger.debug(f"Reading a warm-up frame for {self.serial_number}...")
+            self.read()  # NOTE(Steven): For now we just read one frame, we could also loop for X frames/secs
```
I wasn't thinking of explicitly checking if the frame is dark but having some heuristic as to how much time do we need (from what you observed on the actual cams) until we get a valid image. The reason is I've removed the warmup we had in the control_robot script and I want instead to design the camera such that when they're connected, the frames read are valid (meaning we do warmup during connection). I think we'll also need to add a warmup time in the cam's config to set it for different cameras, my understanding being that this warmup will vary depending on the camera's sensor.

imstevenpmwork replied:
You're correct—the time needed for a camera's sensors and data stream to stabilize can vary across models and systems.

To handle this consistently, I suggest emulating the time-based warmup approach we previously used in `control_robot`. I can add a configurable warmup period where `connect()` only succeeds if the `read()` calls during that time complete successfully.

aliberts replied:
Yes, let's do that and then if `warmup=False` this should be skipped.

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/47d5008407af46a06cf9bfa69b0c7c7471e3ace8

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
````diff
@@ -13,523 +13,660 @@
-    config = RealSenseCameraConfig(serial_number=128422271347, fps=90, width=640, height=480, color_mode="bgr")
-    # Note: might error out upon `camera.connect()` if these settings are not compatible with the camera
-    ```
+        self.logs: dict = {}  # For timestamping or other metadata
````
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-        if this_name_count > 1:
-            # TODO(aliberts): Test this with multiple identical cameras (Aloha)
+    @staticmethod
+    def find_cameras(raise_when_empty: bool = True) -> List[Dict[str, Any]]:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        Args:
+            raise_when_empty (bool): If True, raises an OSError if no cameras are found.
+
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        Raises:
+            OSError: If `raise_when_empty` is True and no cameras are detected,
+                     or if pyrealsense2 is not installed.
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+                raise OSError(
+                    "No RealSense devices detected. Ensure cameras are connected, "
+                    "library (`pyrealsense2`) is installed, and firmware is up-to-date."
+                )
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-                    f"`serial_number` is expected to be one of these available cameras {serial_numbers}, but {self.serial_number} is provided instead. "
-                    "To find the serial number you should use, run `python lerobot/common/robot_devices/cameras/intelrealsense.py`."
-                )
+    def connect(self, do_warmup_read: bool = True):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+            ValueError: If the configuration is invalid (e.g., missing serial/name, name not unique).
+            ConnectionError: If the camera is found but fails to start the pipeline.
+            RuntimeError: If the pipeline starts but fails to apply requested settings.
+            OSError: If no RealSense devices are detected at all.
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+            self.rs_profile = None
+            self.rs_pipeline = None
+            raise ConnectionError(
+                f"Failed to open RealSense camera {self.serial_number}. Error: {e}. "
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+                f"Run 'python -m find_cameras list-cameras' for details."
+            ) from e
+
+        logger.debug(f"Validating stream configuration for {self.serial_number}...")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+            logger.debug(f"Reading a warm-up frame for {self.serial_number}...")
+            self.read()  # NOTE(Steven): For now we just read one frame, we could also loop for X frames/secs
+
+        logger.info(f"Camera {self.serial_number} connected and configured successfully.")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        logger.debug(f"Validating stream configuration for {self.serial_number}...")
+        self._validate_capture_settings()
+
+        if do_warmup_read:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        """Validates and sets the internal FPS based on actual stream FPS."""
+        actual_fps = stream.fps()
+
+        if self.fps is None:
```
Just checked and `rs_pipeline.start()` will indeed raise with an unsupported `fps`. We can then do:
CURRENT:
```python
if self.fps is None:
            self.fps = stream.fps()
else:
            self._validate_fps(stream)
```
NEW:
```python
if self.fps is None:
            self.fps = stream.fps()
```
And then delete `_validate_fps()`. WDYT?

The `if` outside of the function was addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/39f908e9db68e82a8e667a3ff4373dd06c587a99

aliberts replied:
Sounds good, let's do that

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/47d5008407af46a06cf9bfa69b0c7c7471e3ace8

aliberts replied:
Did you check if also need `_validate_width_and_height`?

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/fb3b45b64f4bc4b7aa726235e3562bc8a42f2e7e

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-        self.fps = round(actual_fps)
-        self.capture_width = round(actual_width)
-        self.capture_height = round(actual_height)
+    def read_depth(self, timeout_ms: int = 5000) -> np.ndarray:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-        """Read a frame from the camera returned in the format height x width x channels (e.g. 480 x 640 x 3)
-        of type `np.uint8`, contrarily to the pytorch format which is float channel first.
+        Args:
+            timeout_ms (int): Maximum time in milliseconds to wait for a frame. Defaults to 5000ms.
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        ret, frame = self.rs_pipeline.try_wait_for_frames(
+            timeout_ms=timeout_ms
+        )  # NOTE(Steven): This read has a timeout
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f
However, my point in here was just to bring the attention to the fact that the realsense read has a timeout while the opencv read will block indefinitely

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        if not ret or frame is None:
+            raise RuntimeError(
+                f"Failed to capture frame from {self}. '.read_depth()' returned status={ret} and frame is None."
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        self.logs["timestamp_utc"] = capture_timestamp_utc()
+        return depth_map_processed
+
+    def read(self, color_mode: ColorMode | None = None, timeout_ms: int = 5000) -> np.ndarray:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        from the camera hardware via the RealSense pipeline.
+
+        Args:
+            timeout_ms (int): Maximum time in milliseconds to wait for a frame. Defaults to 5000ms.
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-                self.color_image, self.depth_map = self.read()
-            else:
-                self.color_image = self.read()
+    def _ensure_read_thread_running(self):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-        """Access the latest color image"""
+        self.stop_event = Event()
+        self.thread = Thread(
+            target=self._read_loop, args=(), name=f"RealSenseReadLoop-{self}-{self.serial_number}"
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+        logger.debug(f"Read thread started for {self}.")
+
+    # NOTE(Steven): Missing implementation for depth for now
+    def async_read(self, timeout_ms: float = 2000) -> np.ndarray:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+
+        Args:
+            timeout_ms (float): Maximum time in milliseconds to wait for a frame
+                to become available in the queue. Defaults to 2000ms (2 seconds).
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-    args = parser.parse_args()
-    save_images_from_cameras(**vars(args))
+        logger.debug(f"Disconnecting from camera {self.serial_number}...")
+
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+            self.rs_pipeline = None
+            self.rs_profile = None
+
+        logger.info(f"Camera {self.serial_number} disconnected successfully.")
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/configuration_realsense.py:
```diff
@@ -14,58 +14,74 @@
-    force_hardware_reset: bool = True
-    rotation: int | None = None
-    mock: bool = False
+    rotation: Cv2Rotation = Cv2Rotation.NO_ROTATION  # NOTE(Steven): Check if draccus can parse to an enum
```
I'm not sure I understand this. Could you explain further @aliberts ? I guess if any change is required we will also need to have it for `ColorMode`

aliberts replied:
What I meant:
```diff
- class ColorMode(Enum):
+ class ColorMode(str, Enum):
    RGB = "rgb"
    BGR = "bgr"


- class Cv2Rotation(Enum):
+ class Cv2Rotation(int, Enum):
    NO_ROTATION = 0
    ROTATE_90 = 90
    ROTATE_180 = 180
    ROTATE_270 = -90
```

Also I don't have an opinion on using an Enum for this or not. My main concern being easy serialization/deserialization to json files for the configs

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/47d5008407af46a06cf9bfa69b0c7c7471e3ace8

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
 
+        self.fps: int | None = config.fps
+        self.channels: int = config.channels
+        self.color_mode: ColorMode = config.color_mode
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
 
-def get_camera_index_from_unix_port(port: Path) -> int:
-    return int(str(port.resolve()).removeprefix("/dev/video"))
+        self.videocapture_camera: cv2.VideoCapture | None = None
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
-    path = images_dir / f"camera_{camera_index:02d}_frame_{frame_index:06d}.png"
-    path.parent.mkdir(parents=True, exist_ok=True)
-    img.save(str(path), quality=100)
+        self.logs: dict = {}  # NOTE(Steven): Might be removed in the future
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+        self.logs: dict = {}  # NOTE(Steven): Might be removed in the future
 
+        self.rotation: int | None = get_cv2_rotation(config.rotation)
+        self.backend: int = get_cv2_backend()  # NOTE(Steven): If we specify backend the opencv open fails
```
Observation in Ubuntu 24.04.2 LTS (x86_64):

- When running:  
  ```python  
  self.videocapture = cv2.VideoCapture(self.index_or_path)  # No backend specified  
  ```  
  OpenCV defaults to the `V4L2` backend, and `self.videocapture.isOpened()` returns `True`.  

- However, when explicitly setting the backend to `cv2.CAP_DSHOW` (via `get_cv2_backend()`):  
  ```python  
  self.videocapture = cv2.VideoCapture(self.index_or_path, self.backend)  
  ```  
  `self.videocapture.isOpened()` returns `False`.

Behavior in `macOS` is fine even if we specify the backend (but in this case, it get's resolved to `AVFOUNDATION`).

I wonder if we should just leave the backend resolve to the one that it is found/resolved to by default.

aliberts replied:
These backends were added in #676 to solve an issue on Windows. Let's change the backend for Linux to `cv2.V4L2` in `get_cv2_backend` then.
Also according to the function's code, it should resolve to `cv2.CAP_ANY` on macOS. Did you observe a different behavior?

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/47d5008407af46a06cf9bfa69b0c7c7471e3ace8

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
-            if time.perf_counter() - start_time > record_time_s:
-                break
+    def __str__(self) -> str:
+        """Returns a string representation of the camera instance."""
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-        self.capture_width = config.width
-        self.capture_height = config.height
+    def __str__(self) -> str:
+        """Returns a string representation of the camera instance."""
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
-        self.config = config
-        self.camera_index = config.camera_index
-        self.port = None
+        logger.debug(f"Attempting to connect to camera {self.index_or_path} using backend {self.backend}...")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
-        # Store the raw (capture) resolution from the config.
-        self.capture_width = config.width
-        self.capture_height = config.height
+        logger.debug(f"Successfully opened camera {self.index_or_path}. Applying configuration...")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
-        self.is_connected = True
+    @staticmethod
+    def find_cameras(
+        max_index_search_range=MAX_OPENCV_INDEX, raise_when_empty: bool = True
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+
+        logger.debug(f"Stopping read loop thread for {self}.")
+
+    def _ensure_read_thread_running(self):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+        if not self.is_connected and self.thread is None:
             raise DeviceNotConnectedError(
-                f"OpenCVCamera({self.camera_index}) is not connected. Try running `camera.connect()` first."
+                f"Attempted to disconnect {self}, but it appears already disconnected."
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+                f"Attempted to disconnect {self}, but it appears already disconnected."
             )
 
+        logger.debug(f"Disconnecting from camera {self.index_or_path}...")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+            self.videocapture_camera.release()
+            self.videocapture_camera = None
+
+        logger.info(f"Camera {self.index_or_path} disconnected successfully.")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,541 @@
+            self._shutdown_read_thread()
+
+        if self.videocapture_camera is not None:
+            logger.debug(f"Releasing OpenCV VideoCapture object for {self}.")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/3890c38c12a53cf1a041a773b65b3b4b02e3d7a2

imstevenpmwork commented on lerobot/common/cameras/opencv/configuration_opencv.py:
```diff
@@ -1,38 +1,76 @@
-    mock: bool = False
+    index_or_path: int | Path
+    color_mode: ColorMode = ColorMode.RGB
+    channels: int = 3  # NOTE(Steven): Why is this a config?
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on lerobot/common/utils/control_utils.py:
```diff
@@ -38,6 +38,7 @@
 from lerobot.common.utils.utils import get_safe_torch_device, has_method
 
 
+# NOTE(Steven): Consider integrating this in camera class
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,357 @@
+from lerobot.common.cameras.opencv.configuration_opencv import OpenCVCameraConfig
+
+logger = logging.getLogger(__name__)
+# logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(module)s - %(message)s")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,190 @@
+def test_base_class_implementation():
+    config = OpenCVCameraConfig(index_or_path=0)
+
+    _ = OpenCVCamera(config)
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+BAG_FILE_PATH = os.path.join(TEST_ARTIFACTS_DIR, "test_rs.bag")
+
+# NOTE(Steven): Missing tests for depth
+# NOTE(Steven): Takes 20sec, the patch being the biggest bottleneck
```
Last investigation I carried led to the flame graph I shared. It seems that the patching is time expensive, I will double check if there's anything we can do about it. Potentially using an auto fixture as suggested below

imstevenpmwork commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+# NOTE(Steven): more tests + assertions?
+if not os.path.exists(BAG_FILE_PATH):
+    print(f"Warning: Bag file not found at {BAG_FILE_PATH}. Some tests might fail or be skipped.")
+
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+
+def mock_rs_config_enable_device_from_file(rs_config_instance, sn):
+    if not os.path.exists(BAG_FILE_PATH):
+        raise FileNotFoundError(f"Test bag file not found: {BAG_FILE_PATH}")
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+    return rs_config_instance.enable_device_from_file("non_existent_file.bag", repeat_playback=True)
+
+
+def test_base_class_implementation():
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on tests/cameras/test_realsense.py:
```diff
@@ -0,0 +1,199 @@
+    ],
+)
+@patch("pyrealsense2.config.enable_device", side_effect=mock_rs_config_enable_device_from_file)
+def test_all_rotations(mock_enable_device, rotation):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,190 @@
+        Cv2Rotation.ROTATE_270,
+    ],
+)
+def test_all_rotations(rotation, index_or_path):
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on .gitattributes:
```diff
@@ -11,10 +11,11 @@
 *.arrow filter=lfs diff=lfs merge=lfs -text
 *.json !text !filter !merge !diff
+tests/artifacts/cameras/*.png filter=lfs diff=lfs merge=lfs -text
+tests/artifacts/cameras/*.bag filter=lfs diff=lfs merge=lfs -text
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/f27f5f84b0a94cf80c0881d2a3682039aa60814e

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
+            logger.warning(
+                f"Requested FPS {self.fps} for {self}, but camera reported {actual_fps}. "
+                "This might be due to camera limitations."
+            )
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented on lerobot/common/cameras/intel/camera_realsense.py:
```diff
@@ -13,523 +13,660 @@
-    def connect(self):
-        if self.is_connected:
-            raise DeviceAlreadyConnectedError(f"RealSenseCamera({self.serial_number}) is already connected.")
+    def _configure_realsense_settings(self) -> rs.config:
```
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1f2cfd38280c24de0442f2bb5d6d3edd00bb089f

imstevenpmwork commented:
> Overall good progress but I find there are too many logging statements that are adding noise and not much help. I've added suggestion to remove some of them but I didn't do it for all of them as there were many, let's clean that up a bit.

I went ahead and removed 14 more logging statements in commit: https://github.com/huggingface/lerobot/pull/1108/commits/73bb9709a7a6d4fc9ae983d7cc1bf6efea44199c. This in addition to the deletions suggested in the PR review

aliberts commented on lerobot/common/cameras/realsense/configuration_realsense.py:
```diff
@@ -0,0 +1,80 @@
+            )
+
+        values = (self.fps, self.width, self.height)
+        if any(v is not None for v in values) and any(v is None for v in values):
```
I don't think this is what you want

imstevenpmwork replied:
This if state will be `True` if there's at least one _not_ `None` value **and** at least one `None` value. Isn't this what we are looking for?: Either all of them are set or all of them are `None`, but something in between is no good

aliberts replied:
> Either all of them are set or all of them are None

Yes, this is what we want

> This if state will be True if there's at least one not None value and at least one None value

This is not the same

imstevenpmwork replied:
I'm not sure I understand the difference. I tested this `if` statement logic and this was the result:

```
λ › python lerobot/test.py
(None, None, None)   => Raise? False
(None, None, 1)      => Raise? True
(None, 1, None)      => Raise? True
(None, 1, 1)         => Raise? True
(1, None, None)      => Raise? True
(1, None, 1)         => Raise? True
(1, 1, None)         => Raise? True
(1, 1, 1)            => Raise? False
```

Do you still prefer to go back how it was done before? Happy to implement either way 😊

Note:
This is the small test snippet I used:

```python
from itertools import product


def if_statement(values):
    return any(v is not None for v in values) and any(v is None for v in values)


possible_values = [None, 1]
all_permutations = list(product(possible_values, repeat=3))

for case in all_permutations:
    print(f"{str(case):20} => Raise? {if_statement(case)}")
```

aliberts replied:
my bad, I got confused bc I would have written it un-DeMorgan-ified, smt like
```python
not all(v is not None for v in values) and not all(v is None for v in values)
```
but ok if you prefer this form.

aliberts commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,336 @@
+            logger.info(f"Image capture finished. Images saved to {output_dir}")
+
+
+# NOTE(Steven):
+# * realsense identified as opencv -> consistent in linux, we can even capture images
+# * opencv mac cams reporting different fps at init, not an issue as we don't enforce fps here
+# * opencv not opening in linux if we specify a backend
```
Can you move this Note in the docstring at the top?

aliberts commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,336 @@
+
+
+def save_images_from_all_cameras(
+    output_dir: str | Path,
```
```suggestion
    output_dir: Path,
```

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

aliberts commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,336 @@
+def initialize_output_directory(output_dir: str | Path) -> Path:
+    """Initialize and clean the output directory."""
+    output_dir = Path(output_dir)
+    if output_dir.exists():
+        logger.info(f"Output directory {output_dir} exists. Removing previous content.")
+        shutil.rmtree(output_dir)
```
If we're going to `rm` user input directory, I'd want an `overwrite` option or an `input` prompt for confirmation.

imstevenpmwork replied:
To ensure compatibility with systems that may not have a keyboard, I would prefer to avoid user input. Alternatively, we could append a timestamp to the filename, eliminating the need for deletions since each file would have a unique name. WDYT?

aliberts replied:
Let's simply check if the files exist and remove them if they do (instead of deleting the entire folder)

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/2bf433a6774a0e57cac49418cfa86d5424c0cd17

aliberts commented on tests/cameras/test_opencv.py:
```diff
@@ -0,0 +1,194 @@
+        camera.connect(warmup=False)
+
+
+@pytest.mark.parametrize("index_or_path", TEST_IMAGE_PATHS, ids=[str(path.name) for path in TEST_IMAGE_PATHS])
```
```suggestion
@pytest.mark.parametrize("index_or_path", TEST_IMAGE_PATHS, ids=TEST_IMAGE_SIZES)
```

aliberts reviewed (changes requested):
I am sorry but there are still way too many log statements that I don't find helpful (>90% of my comments). Let's clean this up.

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,501 @@
-                self.color_image = self.read()
+                color_image = self.read()
+
+                with contextlib.suppress(queue.Empty):
+                    _ = self.frame_queue.get_nowait()
```
`self.frame_queue.put(color_image)` could block if the queue is full. To avoid this, we always remove the oldest frame (if one exists) before adding a new one. This ensures the queue always contains the latest frame, even if no one is actively reading from it.

For example, if the loop keeps running without any reads, we still want the most recent frame available for when `async_read` is called. Since the queue only holds one frame, we discard the previous one before adding the next.

aliberts replied:
But then why not use a Lifo queue? This is a hack
I think if we're to have a single queue element let's just use a lock and read outside of it. Something like this:
```python
color_image = self.read()
with self.lock:
    self.frame = color_image
```

aliberts replied:
Between `get_nowait`() and `put`() the queue is empty. If the main thread reads from it right then it won't get anything.

Let's use a lock for this particular case and handle the timeout ourselves (in `async_read`)

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1c6b349dd49fb31730f19fd50fcc593f51b3b3dc

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,501 @@
+                logger.error(f"Read loop for {self} stopped: Camera disconnected.")
+                break
             except Exception as e:
-                print(f"Error reading in thread: {e}")
+                logger.warning(f"Error reading frame in background thread for {self}: {e}")
```
The try/except block ensures robust continuous operation even if temporary errors occur.
1) We handle disconnection errors specifically by breaking the loop
2) We recover from other errors (like temporary camera issues) by just continuing, maintaining operation for clients who might not currently be using async_read()  

Not raising in here provides better stability since brief camera glitches won't crash the application - we simply skip a frame update rather than failing completely. Only permanent disconnection stops the thread.

imstevenpmwork commented on lerobot/common/cameras/realsense/camera_realsense.py:
```diff
@@ -0,0 +1,612 @@
+            except DeviceNotConnectedError:
+                logger.error(f"Read loop for {self} stopped: Camera disconnected.")
+                break
+            except Exception as e:
+                logger.warning(f"Error reading frame in background thread for {self}: {e}")
```
Same as above:
The try/except block ensures robust continuous operation even if temporary errors occur.
1) We handle disconnection errors specifically by breaking the loop
2) We recover from other errors (like temporary camera issues) by just continuing, maintaining operation for clients who might not currently be using async_read()  

Not raising in here provides better stability since brief camera glitches won't crash the application - we simply skip a frame update rather than failing completely. Only permanent disconnection stops the thread.

imstevenpmwork commented on lerobot/common/cameras/utils.py:
```diff
@@ -12,10 +37,39 @@ def make_cameras_from_configs(camera_configs: dict[str, CameraConfig]) -> dict[s
+        return cv2.CAP_ANY
+
+
+def save_image(img_array: np.ndarray, camera_index: int, frame_index: int, images_dir: Path):
+    img = Image.fromarray(img_array)
+    path = images_dir / f"camera_{camera_index:02d}_frame_{frame_index:06d}.png"
+    path.parent.mkdir(parents=True, exist_ok=True)
+    img.save(str(path), quality=100)
```
Here: https://github.com/huggingface/lerobot/blob/bf099932a1197b147f6d95cf27ec1d4dbcf4acf6/lerobot/find_cameras.py#L213

aliberts replied:
No it's not: that's using a function with the same name defined in that file

imstevenpmwork commented on lerobot/common/cameras/camera.py:
```diff
@@ -1,25 +1,122 @@
+        fps (int | None): Configured frames per second
+        width (int | None): Frame width in pixels
+        height (int | None): Frame height in pixels
+        warmup_time (int | None): Time reading frames before returning from connect (in seconds)
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/common/cameras/camera.py:
```diff
@@ -1,25 +1,122 @@
+        self.fps: int | None = config.fps
+        self.width: int | None = config.width
+        self.height: int | None = config.height
+        self.warmup_time: int | None = config.warmup_time
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/common/cameras/configs.py:
```diff
@@ -1,11 +1,45 @@
+    fps: int | None = None
+    width: int | None = None
+    height: int | None = None
+    warmup_time: int | None = None
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,501 @@
+            if self.warmup_time is None:
+                raise ValueError(
+                    f"Warmup time is not set for {self}. Please set a warmup time in the configuration."
+                )
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,501 @@
+        h, w, c = image.shape
 
-        h, w, _ = color_image.shape
         if h != self.capture_height or w != self.capture_width:
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/common/cameras/realsense/camera_realsense.py:
```diff
@@ -0,0 +1,612 @@
+            if self.warmup_time is None:
+                raise ValueError(
+                    f"Warmup time is not set for {self}. Please set a warmup time in the configuration."
+                )
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/common/cameras/realsense/camera_realsense.py:
```diff
@@ -0,0 +1,612 @@
+        else:
+            h, w, _c = image.shape
+
+        if h != self.capture_height or w != self.capture_width:
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/96d9c40eb8010c3271be604f471baa6d7e0867da

imstevenpmwork commented on lerobot/find_cameras.py:
```diff
@@ -0,0 +1,336 @@
+
+def initialize_output_directory(output_dir: str | Path) -> Path:
+    """Initialize and clean the output directory."""
+    output_dir = Path(output_dir)
```
Addressed in commit: https://github.com/huggingface/lerobot/commit/e556c97c08413ff2908dce772888d9a6f0d5424b

aliberts commented on lerobot/common/cameras/realsense/camera_realsense.py:
```diff
@@ -0,0 +1,612 @@
+            try:
+                frame_data = self.read(timeout_ms=500)
+
+                with contextlib.suppress(queue.Empty):
+                    _ = self.frame_queue.get_nowait()
```
(see previous answer)

imstevenpmwork replied:
Same as above

imstevenpmwork replied:
Addressed in commit: https://github.com/huggingface/lerobot/pull/1108/commits/1c6b349dd49fb31730f19fd50fcc593f51b3b3dc

aliberts commented on lerobot/common/cameras/opencv/camera_opencv.py:
````diff
@@ -13,507 +13,473 @@
-    # when done using the camera, consider disconnecting
-    camera.disconnect()
-    ```
+        logger.debug(f"{self} connected.")
````
(for consistency)
```suggestion
        logger.info(f"{self} connected.")
```

aliberts commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,473 @@
 
-        if not ret:
-            raise OSError(f"Can't capture color image from camera {self.camera_index}.")
+        # Post-process the frame (color conversion, dimension check, rotation)
```
not needed
```suggestion
```

aliberts commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,473 @@
+        if self.thread is None or not self.thread.is_alive():
+            self._start_read_thread()
+
+        event_was_set = self.new_frame_event.wait(timeout=timeout_ms / 1000.0)
+
+        if not event_was_set:
```
nit
```suggestion
        if not self.new_frame_event.wait(timeout=timeout_ms / 1000.0):
```

aliberts commented on lerobot/common/cameras/realsense/camera_realsense.py:
```diff
@@ -0,0 +1,560 @@
+        if self.thread is None or not self.thread.is_alive():
+            self._start_read_thread()
+
+        event_was_set = self.new_frame_event.wait(timeout=timeout_ms / 1000.0)
+
+        if not event_was_set:
```
nit
```suggestion
        if not self.new_frame_event.wait(timeout=timeout_ms / 1000.0):
```

aliberts commented on lerobot/common/cameras/opencv/camera_opencv.py:
```diff
@@ -13,507 +13,473 @@
-        num_tries = 0
-        while True:
-            if self.color_image is not None:
-                return self.color_image
+            if frame is None:
+                raise RuntimeError(f"Internal error: Event set but no frame available for {self}.")
 
-            time.sleep(1 / self.fps)
-            num_tries += 1
-            if num_tries > self.fps * 2:
-                raise TimeoutError("Timed out waiting for async_read() to start.")
+            return frame
+        except Exception as e:
+            raise RuntimeError(f"Error getting frame from camera {self}: {e}") from e
```
This weird, let's remove this try/except block as I don't think it's needed

aliberts commented on lerobot/common/cameras/realsense/camera_realsense.py:
```diff
@@ -0,0 +1,560 @@
+                f"Read thread alive: {thread_alive}."
+            )
+
+        try:
+            with self.frame_lock:
+                frame = self.latest_frame
+                self.new_frame_event.clear()
+
+            if frame is None:
+                raise RuntimeError(f"Internal error: Event set but no frame available for {self}.")
+
+            return frame
+        except Exception as e:
+            raise RuntimeError(f"Error getting frame from camera {self}: {e}") from e
```
Same comment as for opencv

aliberts reviewed (changes requested):
Last few changes and we're good to go ;)

aliberts reviewed (approved):
LGTM, thank you!
